Implement Classifier and Base Forwarder Selection Algorithm to Resolve N+1 Query Problem

jhmin99 opened the pull request:
### Description
This pull request introduces the `Classifier` and lays the foundation for the `Forwarder selection algorithm`, addressing the `N+1 query problem` and enhancing the efficiency and scalability of `Peer-to-Peer (P2P)` connections.

#### Key Architectural Changes: 
1. Consolidation of `ForwarderStatusInfo` into `ClientInfo`:   
- Initially, a separate `ForwarderStatusInfo table` was considered, but to avoid synchronization issues and simplify management, all relevant information is integrated into `ClientInfo`.
2. Introduction of a new class hierarchy for clients:
- `Candidate`: Newly connected clients receiving media directly from the server.
- `Potential Forwarder`: Clients capable of forwarding media but require confirmation through classification.

### Related Issues
- #64 

### Changes Made
#### Classifier Implementation:
The Classifier dynamically assigns or updates client classes during specific scenarios:
- Triggered By:
   - `Activate`: Upon new client activation in the system.
   - `Balancing Failure`: When a P2P connection attempt fails.
 - Classification Workflow:
    1. The `coordinator` publishes a classification event to the `Broker` upon activation or balancing failure.
    2. The `Classifier` retrieves Fetchers and Potential Forwarders and pairs them using a `Round-Robin` approach.
    3. `SDP Exchange`:
       1. The server intermediates the SDP exchange between the Fetcher and Potential Forwarder.
       2. Results are published back to the Broker.
     4.  Class Reassignment:
       - `Success`: Potential Forwarder → `Forwarder`
       - `Failure`: Potential Forwarder → `Fetcher`
- Future Scalability:
  - While currently event-driven, periodic classification mechanisms can be added as the system scales.   
   
#### Forwarder Selection Algorithm:
A new scoring mechanism ensures efficient and dynamic `Forwarder selection`.
- Algorithm Details:
  - Differentiates clients based on their class (example):
     - `Candidate`: 0.0
     - `Potential Forwarder`: 3.0   
     - `Forwarder`: 5.0 
  - Additional Metrics (currently implemented):
    - `Connection Count`:
      - `Lower` counts yield `higher` scores.
      - Clients exceeding the maximum allowed connections are excluded.
    - `Created Time`:
      - Older clients are prioritized due to presumed stability and reliability.    
  - Future Metrics (planned in upcoming PRs):
    - `Network Performance`: Prefer clients with better bandwidth and lower latency.
    - `Packet Loss Rate`: Penalize clients with high packet loss.
    - `CPU Usage`: Avoid clients under heavy processing loads. 
-  Future Work: `Background Metric Updates`
   - In the next PR, a background worker will be introduced to `periodically update metrics` in the database. This will include key metrics such as `network performance`, `packet loss rate`, and `CPU usage`.   
   - The current algorithm is already designed to handle these metrics dynamically, so no changes will be needed to the core logic. With regular updates to `ClientInfo`, the system will seamlessly leverage the new metrics for `Forwarder selection`, ensuring consistent and optimal performance.
      
<img width="798" alt="스크린샷 2025-01-02 오후 3 21 31" src="https://github.com/user-attachments/assets/71b5bd07-8a80-402f-b033-6f6c7f0eeaf2" />


### Checklist
- [ ] Tests have been added/updated or not required
- [x] didn't break anything

### Additional Context

### Reviewer Notes
- Review the `Classifier workflow` for clarity and potential `edge cases`.
- Feedback on the `Forwarder scoring system` and `future metrics` would be appreciated.

window9u commented on coordinator/coordinator.go:
```diff
@@ -84,6 +84,13 @@ func (c *Coordinator) handleActivate(event any) {
 		return
 	}
+
+	if err := c.broker.Publish(broker.Classification, broker.CLASSIFY, message.ClassifyRequest{
```
What if activated client was pusher?

How about setting classifier trigger `activated` -> `pull completed`??

window9u commented on coordinator/coordinator.go:
```diff
@@ -296,6 +306,12 @@ func (c *Coordinator) handlePeerFailed(event any) {
 		return
 	}
+	if err := c.broker.Publish(broker.Classification, broker.CLASSIFY, message.ClassifyRequest{
+		ChannelID: connInfo.ChannelID,
```
How about not publish event and subscribe peerFailed event in classifier?? If we using like this way, then it is better to call as function.

jhmin99 replied:
If we subscribe to the peer failed event, the message types received by the broker, such as classifyRequest and failed, are different. This would likely require adding or modifying a separate handleClassification logic. What are your thoughts on this?

window9u commented on database/client_info.go:
```diff
@@ -14,27 +14,57 @@ const (
 	Fetcher
+
+	// PotentialForwarder is the class of a user who might be eligible to become a forwarder.
+	PotentialForwarder
```
I think that it could make confusion `Candidate` and `Potential Forwarder`.
I suggest that

`Candidate` -> `Newbee` or `Init` or etc...
`PotentialForwarder` -> `Candidate`

window9u commented on database/client_info.go:
```diff
@@ -14,27 +14,57 @@ const (
+	Class           int
+	ConnectionCount int
+	CreatedAt       time.Time
+	LastUpdated     time.Time
```
Also, I suggest that add `FetchFrom` in Client Info?
I thought that you've added `Potential Forwarder` because of there is no way to configure client is new commer or candidated or candidating.

So if there is information about fetch from, `Server` or `Client` or ` `, it is more configurable and we don't need to add `Potential Forwarder`.

window9u commented on types/client/request/request.go:
```diff
@@ -67,3 +70,26 @@ type Failed struct {
 }
+
+// ClassifySignal is data type for exchanging SDP while classifying
+type ClassifySignal struct {
```
How about integrate this to `Signal`?? Actually `Pull` and `Forward` is also using same `Signal`. And I think we don't need to send `PeerID` for security.

window9u commented on types/client/request/request.go:
```diff
@@ -67,3 +70,26 @@ type Failed struct {
+}
+
+// ClassifyForward is data type for forwarding while classifying
+type ClassifyForward struct {
```
I think this could also be integrated as `Signal` if server saving `ConnectionID`.

window9u commented on types/client/request/request.go:
```diff
@@ -67,3 +70,26 @@ type Failed struct {
+}
+
+// ClassifyResult is data type for classifying result
+type ClassifyResult struct {
```
How about this naming this to `Classified`?

window9u commented:
How about naming like that?
![image](https://github.com/user-attachments/assets/ad2bf260-d613-4f1f-92c7-c784bdd815ab)

And I think naming `Forwarding` logic to this is better. Because `succeed` or `failed` is confusing event as type increased. 

`as-is`
![image](https://github.com/user-attachments/assets/afcda0a8-cb5c-438b-84d9-e2914aa309bf)

`to-be`
![image](https://github.com/user-attachments/assets/438a3ce0-aa5f-462b-b796-4c5269d16cac)



How do you think about that?

window9u commented on classifier/classifier.go:
```diff
@@ -0,0 +1,127 @@
+// New creates a new instance of Classifier.
+func New(c Config, b *broker.Broker, db database.Database) *Classifier {
+	if c.TimeoutDuration == 0 {
+		c.TimeoutDuration = DefaultTimeoutDuration
```
We already set DefaultTimeDuration in cmd and parsing flags. I think it's better to remove this redundunt code.

jhmin99 commented:
## Changes Made:
- Updated logic to manage `Class-related updates` primarily within the `classifier` to ensure centralized handling.
- Renamed `Candidate` → `Newbie` and `Potential Forwarder` → `Candidate` for better semantic clarity.
- Removed the classification publishing logic from the coordinator. Instead, the classifier now subscribes to events like `peer failed` and `pull completed` to handle classification logic.
- Renamed various Request, Response, and Message types for improved consistency and readability.
<img width="588" alt="스크린샷 2025-01-03 오후 6 52 35" src="https://github.com/user-attachments/assets/9fa4481d-c4ac-4622-8d09-8d2b4c4705bf" />
<img width="585" alt="스크린샷 2025-01-03 오후 6 52 45" src="https://github.com/user-attachments/assets/716fe784-bf9a-498a-818e-29b96463b0d1" />

- Added a new `fetchFrom` field to the `ClientInfo `struct to track the source of fetched streams.

### Note:
When utilizing a score-based approach to find the best forwarder for establishing P2P connections, the resulting stream connections could form a structure like a tree. To optimize this, the maximum tree height can be limited in the future to reduce latency and improve efficiency.
<img width="393" alt="스크린샷 2025-01-03 오후 7 04 38" src="https://github.com/user-attachments/assets/3b8c652f-9dac-4517-92d0-4b6290cca58f" />

window9u commented on classifier/classifier.go:
```diff
@@ -0,0 +1,214 @@
+	}
+
+	currentFetcherIndex := 0
+	for _, candidate := range Candidates {
```
I have a question about this logic. In classifying what I understood is that.

1. Get canidates and fetchers.
2. Classifying with both of two.

I think that this mechanism(make all candidates to be classified) should be done by cron job, not for peer connected, peer failed, etc.
because this could be done concurrently. And could be inefficient.

So, I suggest that we should do two of it.
1. total classifying by cron job.
2. classifying one candidates.

In 2, if MediaConnected occurs, then we could classify only for connected client, not all candidates.

jhmin99 commented:
Thank you for your suggestion. I understand your point about the potential inefficiency of classifying all candidates concurrently during peer-related events like media connection or peer failure. Based on our current logic and goals, here’s how I plan to approach this:

1. `Peer Failed` Events:
When a peer fails, all affected peers are downgraded to `fetchers`. We then classify these `fetchers` with the `candidates` to determine their P2P connectivity and update their classes accordingly.

2. `Media Connected` Events:
When a peer connects to media, they are considered `Newbie` as they haven’t been evaluated for balancing yet. In this case, we classify this single peer with one `fetcher` to test their P2P connectivity. 

3. `Cron Job` for Total Classifying:
To ensure the overall system remains up-to-date, we periodically classify all undecided candidates and fetchers using a cron job. 

Let me know if you have any further feedback or suggestions!

jhmin99 commented:
<img width="247" alt="스크린샷 2025-01-07 오후 2 57 01" src="https://github.com/user-attachments/assets/45e66adb-e8be-413f-98ef-ade9c5731a46" />

During the `balancing` process, the situation depicted in the diagram above can occur. Due to the deactivation of ‘A,’ ‘B’ initiates a pull event, allowing it to receive the stream without issues. However, its child nodes lose their existing stream and fail to receive a new one.

To address this, I implemented a simple rebalancing logic alongside the tree height logic mentioned earlier.

As a result:
With the `rebalancing logic` + `tree height logic`, the reconnect functionality now works seamlessly even when `deactivation` occurs in the middle of the process.
